false negative for protected-access in pylint 2.4.0 (new from 2.3.1)

protected-access violations previously detected by 2.3.1 are not detected by 2.4.0

### Steps to reproduce
Run pylint versus sample.py, which is:
```
$ cat sample.py
'''
Example that demonstrates a protected access incorrectly not flagged
'''

class Object1():
    def __init__(self):
        self._private = 0

    def __str__(self):
        return str(self._private)

class Object2():
    def __init__(self, example):
        self.public = example._private
        self._private = example._private

    def __str__(self):
        return str(self.public) + str(self._private)

def main():
    object1 = Object1()
    print(object1)
    object2 = Object2(object1)
    print(object2)

if __name__ == '__main__':
    main()
    raise SystemExit(0)

```

pylint 2.3.1 detects this:
```
$ pylint --version
pylint 2.3.1
astroid 2.3.0
Python 3.7.3 (default, Apr  3 2019, 19:16:38)
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]]
(CLFSLoad_venv) atz5d8a2db350-linux-1:/jazelenk/src/clfsload.10$ pylint sample.py
************* Module sample
sample.py:14:22: W0212: Access to a protected member _private of a client class (protected-access)
sample.py:15:24: W0212: Access to a protected member _private of a client class (protected-access)

------------------------------------------------------------------
Your code has been rated at 8.95/10 (previous run: 8.95/10, +0.00)
```

pylint 2.4.0 does not:
```
$ pylint --version
pylint 2.4.0
astroid 2.3.0
Python 3.7.3 (default, Apr  3 2019, 19:16:38)
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]]
(CLFSLoad_venv) atz5d8a2db350-linux-1:/jazelenk/src/clfsload.10$ pylint sample.py

-------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 8.95/10, +1.05)
```

I'm not sure if this is related to https://github.com/PyCQA/pylint/pull/2959

@jazelenk thanks for the report.
In fact pylint doesn't emit `protected-access` when the protected attribute is accessed inside special methods as it is the case here. This is to prevent false positive when using for example `__eq__` where you may need to access protected member of the object you compare to.
However if the protected access is not inside special method, then pylint emit the message. For example consider the code below:
```
#pylint: disable=missing-docstring, too-few-public-methods
'''
Example that demonstrates a protected access incorrectly not flagged
'''

class Object1():
    def __init__(self):
        self._private = 0

    def __str__(self):
        return str(self._private)

class Object2():
    def __init__(self, example):
        self.public = example._private
        self._private = example._private

    def __str__(self):
        return str(self.public) + str(self._private)

    def _fake_special(self, example):
        self.public = example._private

def main():
    object1 = Object1()
    print(object1)
    object2 = Object2(object1)
    print(object2)

if __name__ == '__main__':
    main()
    raise SystemExit(0)
``` 
the when we run pylint against it we got :
```
bug_pylint_3120.py:22:22: W0212: Access to a protected member _private of a client class (protected-access)
```

As it seems there is no issue here, i close this issue. If i'm wrong or if those explanations are not clear feel free to reopen it.

--- a/ChangeLog
+++ b/ChangeLog
@@ -4,13 +4,16 @@ Pylint's ChangeLog
 
 What's New in Pylint 2.6.1?
 ===========================
-Release date: 2020-09-08
+Release date: TBA
+
+* Adds option ``check-protected-access-in-special-methods`` in the ClassChecker to activate/deactivate
+  ``protected-access`` message emission for single underscore prefixed attribute in special methods.
+
+  Close #3120
 
 * Fix vulnerable regular expressions in ``pyreverse``
 
   Close #3811
-
-Release date: TBA
 
 * ``inconsistent-return-statements`` message is now emitted if one of ``try/except`` statement
   is not returning explicitly while the other do.

--- a/doc/whatsnew/2.6.rst
+++ b/doc/whatsnew/2.6.rst
@@ -24,6 +24,9 @@ New checkers
 Other Changes
 =============
 
+* Adds option ``check-protected-access-in-special-methods`` in the ClassChecker to activate/deactivate
+  ``protected-access`` message emission for single underscore prefixed attribute in special methods.
+
 * ``inconsistent-return-statements`` message is now emitted if one of ``try/except`` statement
   is not returning explicitly while the other do.
 

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -756,6 +756,15 @@ a metaclass class method.",
                 ),
             },
         ),
+        (
+            "check-protected-access-in-special-methods",
+            {
+                "default": False,
+                "type": "yn",
+                "metavar": "<y or n>",
+                "help": "Warn about protected attribute access inside special methods",
+            },
+        ),
     )
 
     def __init__(self, linter=None):
@@ -1418,11 +1427,12 @@ a metaclass class method.",
                     if _is_attribute_property(name, klass):
                         return
 
-                # A licit use of protected member is inside a special method
-                if not attrname.startswith(
-                    "__"
-                ) and self._is_called_inside_special_method(node):
-                    return
+                if not self.config.check_protected_access_in_special_methods:
+                    # A licit use of protected member is inside a special method
+                    if not attrname.startswith(
+                        "__"
+                    ) and self._is_called_inside_special_method(node):
+                        return
 
                 self.add_message("protected-access", node=node, args=attrname)
 

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -1427,12 +1427,14 @@ a metaclass class method.",
                     if _is_attribute_property(name, klass):
                         return
 
-                if not self.config.check_protected_access_in_special_methods:
+                if (
+                    not self.config.check_protected_access_in_special_methods
+                    and
                     # A licit use of protected member is inside a special method
-                    if not attrname.startswith(
-                        "__"
-                    ) and self._is_called_inside_special_method(node):
-                        return
+                    not attrname.startswith("__")
+                    and self._is_called_inside_special_method(node)
+                ):
+                    return
 
                 self.add_message("protected-access", node=node, args=attrname)
 

--- a/tests/checkers/unittest_classes.py
+++ b/tests/checkers/unittest_classes.py
@@ -142,11 +142,17 @@ class TestVariablesChecker(CheckerTestCase):
         )
         classdef = node.body[-1]
         assign_attribute_in_eq = classdef.instance_attr("_protected")[-1]
-        attribute_in_eq = assign_attribute_in_eq.assigned_stmts()[-1]
+        attribute_in_eq = [_ for _ in assign_attribute_in_eq.assigned_stmts()][
+            -1
+        ]  # pylint:disable=unnecessary-comprehension
         assign_attribute_in_fake_1 = classdef.instance_attr("public")[-1]
-        attribute_in_fake_1 = assign_attribute_in_fake_1.assigned_stmts()[-1]
+        attribute_in_fake_1 = [_ for _ in assign_attribute_in_fake_1.assigned_stmts()][
+            -1
+        ]  # pylint:disable=unnecessary-comprehension
         assign_attribute_in_fake_2 = classdef.instance_attr("__private")[-1]
-        attribute_in_fake_2 = assign_attribute_in_fake_2.assigned_stmts()[-1]
+        attribute_in_fake_2 = [_ for _ in assign_attribute_in_fake_2.assigned_stmts()][
+            -1
+        ]  # pylint:disable=unnecessary-comprehension
         with self.assertAddsMessages(
             Message("protected-access", node=attribute_in_eq, args="_protected"),
             Message("protected-access", node=attribute_in_fake_1, args="_protected"),
@@ -179,9 +185,13 @@ class TestVariablesChecker(CheckerTestCase):
         )
         classdef = node.body[-1]
         assign_attribute_in_fake_1 = classdef.instance_attr("public")[-1]
-        attribute_in_fake_1 = assign_attribute_in_fake_1.assigned_stmts()[-1]
+        attribute_in_fake_1 = [_ for _ in assign_attribute_in_fake_1.assigned_stmts()][
+            -1
+        ]  # pylint:disable=unnecessary-comprehension
         assign_attribute_in_fake_2 = classdef.instance_attr("__private")[-1]
-        attribute_in_fake_2 = assign_attribute_in_fake_2.assigned_stmts()[-1]
+        attribute_in_fake_2 = [_ for _ in assign_attribute_in_fake_2.assigned_stmts()][
+            -1
+        ]  # pylint:disable=unnecessary-comprehension
         with self.assertAddsMessages(
             Message("protected-access", node=attribute_in_fake_1, args="_protected"),
             Message("protected-access", node=attribute_in_fake_2, args="__private"),

--- a/tests/checkers/unittest_classes.py
+++ b/tests/checkers/unittest_classes.py
@@ -142,15 +142,11 @@ class TestVariablesChecker(CheckerTestCase):
         )
         classdef = node.body[-1]
         assign_attribute_in_eq = classdef.instance_attr("_protected")[-1]
-        attribute_in_eq = [_ for _ in assign_attribute_in_eq.assigned_stmts()][-1]
+        attribute_in_eq = assign_attribute_in_eq.assigned_stmts()[-1]
         assign_attribute_in_fake_1 = classdef.instance_attr("public")[-1]
-        attribute_in_fake_1 = [_ for _ in assign_attribute_in_fake_1.assigned_stmts()][
-            -1
-        ]
+        attribute_in_fake_1 = assign_attribute_in_fake_1.assigned_stmts()[-1]
         assign_attribute_in_fake_2 = classdef.instance_attr("__private")[-1]
-        attribute_in_fake_2 = [_ for _ in assign_attribute_in_fake_2.assigned_stmts()][
-            -1
-        ]
+        attribute_in_fake_2 = assign_attribute_in_fake_2.assigned_stmts()[-1]
         with self.assertAddsMessages(
             Message("protected-access", node=attribute_in_eq, args="_protected"),
             Message("protected-access", node=attribute_in_fake_1, args="_protected"),
@@ -182,16 +178,10 @@ class TestVariablesChecker(CheckerTestCase):
         """
         )
         classdef = node.body[-1]
-        assign_attribute_in_eq = classdef.instance_attr("_protected")[-1]
-        attribute_in_eq = [_ for _ in assign_attribute_in_eq.assigned_stmts()][-1]
         assign_attribute_in_fake_1 = classdef.instance_attr("public")[-1]
-        attribute_in_fake_1 = [_ for _ in assign_attribute_in_fake_1.assigned_stmts()][
-            -1
-        ]
+        attribute_in_fake_1 = assign_attribute_in_fake_1.assigned_stmts()[-1]
         assign_attribute_in_fake_2 = classdef.instance_attr("__private")[-1]
-        attribute_in_fake_2 = [_ for _ in assign_attribute_in_fake_2.assigned_stmts()][
-            -1
-        ]
+        attribute_in_fake_2 = assign_attribute_in_fake_2.assigned_stmts()[-1]
         with self.assertAddsMessages(
             Message("protected-access", node=attribute_in_fake_1, args="_protected"),
             Message("protected-access", node=attribute_in_fake_2, args="__private"),

--- a/tests/checkers/unittest_classes.py
+++ b/tests/checkers/unittest_classes.py
@@ -141,16 +141,20 @@ class TestVariablesChecker(CheckerTestCase):
         """
         )
         classdef = node.body[-1]
-        assign_attribute_in_eq = classdef.instance_attr('_protected')[-1]
-        attribute_in_eq = [_ for _ in assign_attribute_in_eq.assigned_stmts()][-1] 
-        assign_attribute_in_fake_1 = classdef.instance_attr('public')[-1]
-        attribute_in_fake_1 = [_ for _ in assign_attribute_in_fake_1.assigned_stmts()][-1] 
-        assign_attribute_in_fake_2 = classdef.instance_attr('__private')[-1]
-        attribute_in_fake_2 = [_ for _ in assign_attribute_in_fake_2.assigned_stmts()][-1] 
+        assign_attribute_in_eq = classdef.instance_attr("_protected")[-1]
+        attribute_in_eq = [_ for _ in assign_attribute_in_eq.assigned_stmts()][-1]
+        assign_attribute_in_fake_1 = classdef.instance_attr("public")[-1]
+        attribute_in_fake_1 = [_ for _ in assign_attribute_in_fake_1.assigned_stmts()][
+            -1
+        ]
+        assign_attribute_in_fake_2 = classdef.instance_attr("__private")[-1]
+        attribute_in_fake_2 = [_ for _ in assign_attribute_in_fake_2.assigned_stmts()][
+            -1
+        ]
         with self.assertAddsMessages(
             Message("protected-access", node=attribute_in_eq, args="_protected"),
             Message("protected-access", node=attribute_in_fake_1, args="_protected"),
-            Message("protected-access", node=attribute_in_fake_2, args="__private")
+            Message("protected-access", node=attribute_in_fake_2, args="__private"),
         ):
             self.walk(node.root())
 
@@ -178,14 +182,18 @@ class TestVariablesChecker(CheckerTestCase):
         """
         )
         classdef = node.body[-1]
-        assign_attribute_in_eq = classdef.instance_attr('_protected')[-1]
-        attribute_in_eq = [_ for _ in assign_attribute_in_eq.assigned_stmts()][-1] 
-        assign_attribute_in_fake_1 = classdef.instance_attr('public')[-1]
-        attribute_in_fake_1 = [_ for _ in assign_attribute_in_fake_1.assigned_stmts()][-1] 
-        assign_attribute_in_fake_2 = classdef.instance_attr('__private')[-1]
-        attribute_in_fake_2 = [_ for _ in assign_attribute_in_fake_2.assigned_stmts()][-1] 
+        assign_attribute_in_eq = classdef.instance_attr("_protected")[-1]
+        attribute_in_eq = [_ for _ in assign_attribute_in_eq.assigned_stmts()][-1]
+        assign_attribute_in_fake_1 = classdef.instance_attr("public")[-1]
+        attribute_in_fake_1 = [_ for _ in assign_attribute_in_fake_1.assigned_stmts()][
+            -1
+        ]
+        assign_attribute_in_fake_2 = classdef.instance_attr("__private")[-1]
+        attribute_in_fake_2 = [_ for _ in assign_attribute_in_fake_2.assigned_stmts()][
+            -1
+        ]
         with self.assertAddsMessages(
             Message("protected-access", node=attribute_in_fake_1, args="_protected"),
-            Message("protected-access", node=attribute_in_fake_2, args="__private")
+            Message("protected-access", node=attribute_in_fake_2, args="__private"),
         ):
             self.walk(node.root())

--- a/tests/checkers/unittest_classes.py
+++ b/tests/checkers/unittest_classes.py
@@ -116,3 +116,76 @@ class TestVariablesChecker(CheckerTestCase):
             Message("protected-access", node=node.value, args="_nargs")
         ):
             self.checker.visit_attribute(node.value)
+
+    @set_config(check_protected_access_in_special_methods=True)
+    def test_check_protected_access_in_special_methods(self):
+        """Test that check-protected-access-in-special-methods can be used to
+        trigger protected-access message emission for single underscore prefixed names
+        inside special methods
+        """
+
+        node = astroid.parse(
+            """
+        class Protected:
+            '''empty'''
+            def __init__(self):
+                self._protected = 42
+                self.public = "A"
+                self.__private = None
+            def __eq__(self, other):
+                self._protected = other._protected
+            def _fake_special_(self, other):
+                a = other.public
+                self.public = other._protected
+                self.__private = other.__private
+        """
+        )
+        classdef = node.body[-1]
+        assign_attribute_in_eq = classdef.instance_attr('_protected')[-1]
+        attribute_in_eq = [_ for _ in assign_attribute_in_eq.assigned_stmts()][-1] 
+        assign_attribute_in_fake_1 = classdef.instance_attr('public')[-1]
+        attribute_in_fake_1 = [_ for _ in assign_attribute_in_fake_1.assigned_stmts()][-1] 
+        assign_attribute_in_fake_2 = classdef.instance_attr('__private')[-1]
+        attribute_in_fake_2 = [_ for _ in assign_attribute_in_fake_2.assigned_stmts()][-1] 
+        with self.assertAddsMessages(
+            Message("protected-access", node=attribute_in_eq, args="_protected"),
+            Message("protected-access", node=attribute_in_fake_1, args="_protected"),
+            Message("protected-access", node=attribute_in_fake_2, args="__private")
+        ):
+            self.walk(node.root())
+
+    @set_config(check_protected_access_in_special_methods=False)
+    def test_check_protected_access_in_special_methods_deact(self):
+        """Test that when check-protected-access-in-special-methods is False (default)
+        no protected-access message emission for single underscore prefixed names
+        inside special methods occur
+        """
+
+        node = astroid.parse(
+            """
+        class Protected:
+            '''empty'''
+            def __init__(self):
+                self._protected = 42
+                self.public = "A"
+                self.__private = None
+            def __eq__(self, other):
+                self._protected = other._protected
+            def _fake_special_(self, other):
+                a = other.public
+                self.public = other._protected
+                self.__private = other.__private
+        """
+        )
+        classdef = node.body[-1]
+        assign_attribute_in_eq = classdef.instance_attr('_protected')[-1]
+        attribute_in_eq = [_ for _ in assign_attribute_in_eq.assigned_stmts()][-1] 
+        assign_attribute_in_fake_1 = classdef.instance_attr('public')[-1]
+        attribute_in_fake_1 = [_ for _ in assign_attribute_in_fake_1.assigned_stmts()][-1] 
+        assign_attribute_in_fake_2 = classdef.instance_attr('__private')[-1]
+        attribute_in_fake_2 = [_ for _ in assign_attribute_in_fake_2.assigned_stmts()][-1] 
+        with self.assertAddsMessages(
+            Message("protected-access", node=attribute_in_fake_1, args="_protected"),
+            Message("protected-access", node=attribute_in_fake_2, args="__private")
+        ):
+            self.walk(node.root())

--- a/tests/checkers/unittest_classes.py
+++ b/tests/checkers/unittest_classes.py
@@ -142,17 +142,11 @@ class TestVariablesChecker(CheckerTestCase):
         )
         classdef = node.body[-1]
         assign_attribute_in_eq = classdef.instance_attr("_protected")[-1]
-        attribute_in_eq = [_ for _ in assign_attribute_in_eq.assigned_stmts()][
-            -1
-        ]  # pylint:disable=unnecessary-comprehension
+        attribute_in_eq = list(assign_attribute_in_eq.assigned_stmts())[-1]
         assign_attribute_in_fake_1 = classdef.instance_attr("public")[-1]
-        attribute_in_fake_1 = [_ for _ in assign_attribute_in_fake_1.assigned_stmts()][
-            -1
-        ]  # pylint:disable=unnecessary-comprehension
+        attribute_in_fake_1 = list(assign_attribute_in_fake_1.assigned_stmts())[-1]
         assign_attribute_in_fake_2 = classdef.instance_attr("__private")[-1]
-        attribute_in_fake_2 = [_ for _ in assign_attribute_in_fake_2.assigned_stmts()][
-            -1
-        ]  # pylint:disable=unnecessary-comprehension
+        attribute_in_fake_2 = list(assign_attribute_in_fake_2.assigned_stmts())[-1]
         with self.assertAddsMessages(
             Message("protected-access", node=attribute_in_eq, args="_protected"),
             Message("protected-access", node=attribute_in_fake_1, args="_protected"),
@@ -185,13 +179,9 @@ class TestVariablesChecker(CheckerTestCase):
         )
         classdef = node.body[-1]
         assign_attribute_in_fake_1 = classdef.instance_attr("public")[-1]
-        attribute_in_fake_1 = [_ for _ in assign_attribute_in_fake_1.assigned_stmts()][
-            -1
-        ]  # pylint:disable=unnecessary-comprehension
+        attribute_in_fake_1 = list(assign_attribute_in_fake_1.assigned_stmts())[-1]
         assign_attribute_in_fake_2 = classdef.instance_attr("__private")[-1]
-        attribute_in_fake_2 = [_ for _ in assign_attribute_in_fake_2.assigned_stmts()][
-            -1
-        ]  # pylint:disable=unnecessary-comprehension
+        attribute_in_fake_2 = list(assign_attribute_in_fake_2.assigned_stmts())[-1]
         with self.assertAddsMessages(
             Message("protected-access", node=attribute_in_fake_1, args="_protected"),
             Message("protected-access", node=attribute_in_fake_2, args="__private"),

